frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java

```diff
@@ -46,185 +45,217 @@
-      LOGGER.logWarning(
-          "select: Cannot found a SelectionResult satisfies the Request: " +
-              "Check whether the Request can be relaxed to RM");
-      String notRelaxLogPrefix = "select: The Request cannot be relaxed to RM: Reason: ";
-
-      // Test NodeGpuType
-      if (requestNodeGpuType != null) {
-        throw new NotAvailableException(
-            String.format(notRelaxLogPrefix +
-                    "NodeGpuType is specified: Request NodeGpuType: [%s]",
-                requestNodeGpuType));
+  public synchronized SelectionResult select(ResourceDescriptor requestResource, String taskRoleName) throws NotAvailableException {
+
+    LOGGER.logInfo(
+        "select: TaskRole: [%s] Resource: [%s]", taskRoleName, requestResource);
+    String requestNodeLabel = am.getRequestManager().getTaskPlatParams().get(taskRoleName).getTaskNodeLabel();
+    String requestNodeGpuType = am.getRequestManager().getTaskPlatParams().get(taskRoleName).getTaskNodeGpuType();
+    int pendingTaskNumber = am.getStatusManager().getUnAssociatedTaskCount(taskRoleName);
+    List<ValueRange> reUsePorts = null;
+
+    // Prefer to use previous successfully allocated ports. if no successfully ports, try to re-use the "Requesting" ports.
+    if (am.getRequestManager().getTaskRoles().get(taskRoleName).getUseTheSamePorts()) {
+      reUsePorts = am.getStatusManager().getLiveAssociatedContainerPorts(taskRoleName);
+      if (ValueRangeUtils.getValueNumber(reUsePorts) <= 0 && requestedPortsCache.containsKey(taskRoleName)) {
+        reUsePorts = requestedPortsCache.get(taskRoleName);
+        // the cache only guide the next task to use previous requesting port.
+        requestedPortsCache.remove(taskRoleName);
       }
+    }
+    SelectionResult result = select(requestResource, requestNodeLabel, requestNodeGpuType, pendingTaskNumber, reUsePorts);
+    if (!requestedPortsCache.containsKey(taskRoleName) && pendingTaskNumber > 1)
+      requestedPortsCache.put(taskRoleName, result.getOptimizedResource().getPortRanges());
+    return result;
+  }
 
-      LOGGER.logWarning(
-          "select: The Request will be relaxed to RM");
+  @VisibleForTesting
+  public synchronized SelectionResult select(ResourceDescriptor requestResource, String requestNodeLabel, String requestNodeGpuType, int pendingTaskNumber) throws NotAvailableException {
+    return select(requestResource, requestNodeLabel, requestNodeGpuType, pendingTaskNumber, null);
+  }
+
+  public synchronized SelectionResult select(ResourceDescriptor requestResource, String requestNodeLabel, String requestNodeGpuType,
+      int pendingTaskNumber, List<ValueRange> reUsePorts) throws NotAvailableException {
+
+    LOGGER.logInfo(
+        "select: Request: Resource: [%s], NodeLabel: [%s], NodeGpuType: [%s], TaskNumber: [%d], ReUsePorts: [%s]",
+        requestResource, requestNodeLabel, requestNodeGpuType, pendingTaskNumber, ValueRangeUtils.toString(reUsePorts));
+
+    randomizeNodes();
```

Comment:
>randomizeNodes [](start = 4, length = 14)

if already randomized here, why do you randomize again at the end of setupContainerRequest() in applicationMaster.java?

Reply:
This randomized here is our discussion result, to randomize the selection process,  and this random will also help to random pick Rack. 

The random at the end of setupContainerRequest, is random the final small set of result. because there are some Scoring logic in the selection like packing score,  The final candidates order list usually is the same, we need random pick one in the small set. 

---
In reply to: [173687266](https://github.com/Microsoft/pai/pull/318#discussion_r173687266) [](ancestors = 173687266)